### ISSUE TITLE
Task composition (ie: Celery Canvas)

### ISSUE THREAD

**Issue by aequitas** [Bogdanp/dramatiq#29]
What is your viewpoint and this projects direction in regards to task composition?

I currently use Celery because it allows me to work with tasks on a high level (chain, group, etc). I know Dramatiq supports callbacks but for as far I can see they are rather limited. 

Currently with Dramatiq I cannot trivially:

- Chain more than 2 Actors.
- Pass arguments to a chained Actor during composition (unless I pass them through the first task).
- Composite a more complex task/group and pass it to some other part of my application to have it 'send' (like Celery signatures).

I understand this project might be in it's infancy and you might not want to introduce more complexity then needed, but I'm curious about your roadmap for this project and the use cases.

**Comment by pirate** [comment]
Chaining actors would require adding in a whole event loop system that keeps a queue of tasks that are blocked on a preceding task.

While I see the usefulness, personally I think a scheduler event loop is too complex and high-level to warrant being included in dramatiq.  One of the best things about dramatiq is that it's simple and bare-bones, and if you want to build a more complex system on top of it, there's nothing stopping you.  The requirements for chaining tasks will be so different for different users of this library, I don't see how it can be trivially implemented in a way that suits everyone.  There are so many edge cases around failing tasks, task order dependency, distributing it across multiple workers, what to do when dramatiq goes down halfway through a chained task.

Of course it's up to @Bogdanp, if he thinks there's a simple way to do this safely, I'm all ears.

**Comment by Bogdanp** [comment]
@aequitas can you give a few examples of your use cases? As concrete as possible would be helpful. IME, celery's canvas features end up being a maintenance burden long term and they're generally not very stable (depending on the backend -- I've talked about this elsewhere, but my experience with them has been pretty poor, esp. wrt chord performance and dropped chords).

I agree with @pirate that something like this wouldn't belong in core and can definitely be implemented as an external package.

**Comment by aequitas** [comment]
Thanks for the quick response, just to emphasise I agree with keeping the complexity low. Since you do not want to end up with a second Celery, that'd be pointless. The benefits that Celery brings also has its downsides so it's all about being able to choosing the right tool for the right job.

I think I saw someone on HN mention something like 'Django is to Flask, like Celery is to Dramatiq'. Which is exactly the feeling I had when discovering this project. In that you focus on doing one thing very well and by that produce a clean and lean solution.

That said currently Dramatiq for me is seems to offer to little of a framework for most projects. It does a great job at abstracting an generalising backends and tasks. But it offers little abstraction on top of that so you still have a lot of unrelated stuff you need to account for in your application code making functions/tasks less pure, clean and portable.

I also agree with the suggestion from @pirate that this could be an external package and that it will be hard to fit every use case. Although I think maybe with the right set of primitives you could make it adaptable for a lot of use cases.

As for examples:

https://gitlab.com/failmap/failmap/blob/master/failmap/scanners/scanner_security_headers.py#L86

Here I want to pass the result of the first task to the second task but also provide arguments for the second task. The tasks need to be separated due to access restrictions for the workers. 
The grouping of the tasks in this instance is pure for practicality and is something I would be able to refactor. But it is nice to have a way to handle a bunch of tasks as one entity (ie: groupresult) and interact with that (check overall job completion status, etc). 

My opinion might be tainted a little bit because I've adopted to working the 'Celery way'. I have not formed any vision on how I would like to see tasks queues work outside of the Celery framework. So I might be overthinking things here as well and not see the obvious solutions. But I'm also curious of your experience and vision regarding Dramatiq and how it could cater to a higher abstraction level of development.

**Comment by ericgj** [comment]
I wrote a [middleware](https://gist.github.com/ericgj/6491240e2c132c188f29f5f5ac8dd8a0) for sequencing/chaining tasks -- "sequencing" meaning running each in sequence unless there's a failure, and "chaining" meaning also passing the result of each task to the next.  It also allows you to pass in an "environ" from the top that gets passed in to each task.

Having done it however, I realized there may be a more "integrated" interface. Let's say Actor exposes `message` and `message_with_options` methods.  Then you could (with the support of middleware) do something like

    broker.enqueue(  
        chain( [actor_a.message(*some_args), actor_b.message(*other_args)],
            result_key="_last_result_"
        ) 
    )

It seems like messages could play a role similar to Celery signatures, although I'm not that familiar with Celery.

**Comment by Bogdanp** [comment]
Apologies for the delayed response, it's been a bit of a hectic week for me.

> Having done it however, I realized there may be a more "integrated" interface. Let's say Actor exposes message and message_with_options methods.

I played around with this idea this morning and came up with this:

```python

class Chains(Middleware):
    @property
    def actor_options(self):
        return {"next_in_chain"}

    def after_process_message(self, broker, message, *, result=None, exception=None):
        if exception is not None:
            return

        next_in_chain = message.options.get("next_in_chain")
        if next_in_chain is not None:
            next_message = Message(**next_in_chain)
            broker.enqueue(next_message.copy(args=(result,) + next_message.args))


class chain:
    def __init__(self, children, *, broker=None):
        self.broker = broker or dramatiq.get_broker()
        self.messages = messages = []
        for child in children:
            if isinstance(child, chain):
                messages.extend(message.copy() for message in child.messages)
            else:
                messages.append(child.copy())

        for message, next_in_chain in zip(messages, messages[1:]):
            message.options["next_in_chain"] = next_in_chain.asdict()

    @property
    def first_message(self):
        return self.messages[0]

    @property
    def last_message(self):
        return self.messages[-1]

    def enqueue(self):
        self.broker.enqueue(self.first_message)
        return self

    def get_result(self, *, block: bool=False, timeout: int=None):
        return self.last_message.get_result(block=block, timeout=timeout)


class fanout:
    def __init__(self, children, *, broker=None):
        self.children = list(children)
        self.broker = broker or dramatiq.get_broker()

    def enqueue(self):
        for child in self.children:
            if isinstance(child, (chain, fanout)):
                child.enqueue()
            else:
                self.broker.enqueue(child)
        return self

    def get_result(self, *, block: bool=False, timeout: int=None):
        deadline = None
        if timeout:
            deadline = time.monotonic() + timeout / 1000

        results = []
        for child in self.children:
            if deadline:
                timeout = round(deadline - time.monotonic()) * 1000

            result = child.get_result(block=block, timeout=timeout)
            results.append(result)

        return results
```

Which works pretty well with my little test case:

```python
@dramatiq.actor(store_results=True)
def sleep_then_add(t, x, y):
    time.sleep(t)
    return x + y

if __name__ == "__main__":
    f = fanout([
        fanout([
            sleep_then_add.message(1, 1, 1),
            sleep_then_add.message(1, 1, 2),
            chain([
                sleep_then_add.message(1, 1, 1),
                sleep_then_add.message(1, 2),
                chain([
                    sleep_then_add.message(1, 3),
                    sleep_then_add.message(1, 4),
                ]),
            ]),
        ]),
        sleep_then_add.message(2, 1, 2),
    ]).enqueue()

    for res in f.get_result(block=True):
        print(res)
```

```
[2018-01-13 12:28:27,952] [PID 70694] [Thread-5] [example.sleep_then_add] [INFO] Received args=(1, 1, 2) kwargs={}.
[2018-01-13 12:28:27,954] [PID 70695] [Thread-5] [example.sleep_then_add] [INFO] Received args=(1, 1, 1) kwargs={}.
[2018-01-13 12:28:27,957] [PID 70693] [Thread-5] [example.sleep_then_add] [INFO] Received args=(2, 1, 2) kwargs={}.
[2018-01-13 12:28:27,959] [PID 70689] [Thread-5] [example.sleep_then_add] [INFO] Received args=(1, 1, 1) kwargs={}.
[2018-01-13 12:28:28,957] [PID 70695] [Thread-5] [example.sleep_then_add] [INFO] Completed after 1003.13ms.
[2018-01-13 12:28:28,957] [PID 70694] [Thread-5] [example.sleep_then_add] [INFO] Completed after 1005.21ms.
[2018-01-13 12:28:28,960] [PID 70689] [Thread-5] [example.sleep_then_add] [INFO] Completed after 1000.03ms.
[2018-01-13 12:28:28,988] [PID 70692] [Thread-5] [example.sleep_then_add] [INFO] Received args=(2, 1, 2) kwargs={}.
[2018-01-13 12:28:29,961] [PID 70693] [Thread-5] [example.sleep_then_add] [INFO] Completed after 2004.11ms.
[2018-01-13 12:28:30,990] [PID 70692] [Thread-5] [example.sleep_then_add] [INFO] Completed after 2001.73ms.
[2018-01-13 12:28:31,011] [PID 70690] [Thread-5] [example.sleep_then_add] [INFO] Received args=(3, 1, 3) kwargs={}.
[2018-01-13 12:28:34,015] [PID 70690] [Thread-5] [example.sleep_then_add] [INFO] Completed after 3003.46ms.
[2018-01-13 12:28:34,029] [PID 70691] [Thread-6] [example.sleep_then_add] [INFO] Received args=(4, 1, 4) kwargs={}.
[2018-01-13 12:28:38,030] [PID 70691] [Thread-6] [example.sleep_then_add] [INFO] Completed after 4001.35ms.
```

So it seems like it would be enough for the core package to expose a way to build messages w/o enqueueing them for users to be able to build higher level abstractions. How simple this all turned out to be (API improvements to `chord` and `fanout` notwithstanding) has made me reconsider potentially including these abstractions in the core package, as that would provide a more integrated experience and be less of a hassle both to maintain and set up. I'm going to take some time to think about error handling and other stuff I may be currently missing before I make a decision.

@aequitas @ericgj what do you think of this approach?

**Comment by ericgj** [comment]
Looks great.  Some thoughts:

- The "partial application" style makes a lot of sense to me, and is less of a headache than merging the result in to kwargs, but in the middleware should it be  `broker.enqueue(next_message.copy(next_message.args + args=(result,) ))`   -- putting the result at the end of the fixed args?

- Re error handling, what about specifying a single `on_failure` when you call `chain` or `fanout`, that just gets attached to each child message options -- but otherwise don't do error handling at all?  Then people could make use of the Callbacks middleware or write their own that picks up on `on_failure`.  

- Or to make it more generic: when users call `chain` or `fanout`, they can specify any extra options to get attached to each of their children.

- I'm not particular about names, but should "fanout" be something like "collect"?  To me "fanout" sounds like it describes how tasks are run (in parallel, etc.) rather than how task results are returned, but I may not be familiar enough with common uses of the term "fanout".

**Comment by Bogdanp** [comment]
> The "partial application" style makes a lot of sense to me, and is less of a headache than merging the result in to kwargs, but in the middleware should it be `broker.enqueue(next_message.copy(args=next_message.args + (result,)))` -- putting the result at the end of the fixed args?

Yep, that's what I ended up doing on my branch. It feels a lot more natural than the other way around.

> Re error handling, what about specifying a single on_failure when you call chain or fanout, that just gets attached to each child message options -- but otherwise don't do error handling at all? Then people could make use of the Callbacks middleware or write their own that picks up on on_failure.

> Or to make it more generic: when users call chain or fanout, they can specify any extra options to get attached to each of their children.

Right now I'm leaning towards not adding any and seeing what happens/how it gets used in the wild. `message_with_options` lets you specify failure and success callbacks so maybe that's enough.

> I'm not particular about names, but should "fanout" be something like "collect"? To me "fanout" sounds like it describes how tasks are run (in parallel, etc.) rather than how task results are returned, but I may not be familiar enough with common uses of the term "fanout".

I ended up going with `group` on my branch since fanout felt wrong for the same reasons you describe here. I also renamed `chain` to `pipeline` since I feel like that better describes the concept.

Feel free to leave questions/comments/concerns on https://github.com/Bogdanp/dramatiq/pull/31 .

**Comment by ericgj** [comment]
All sounds good to me!  :100:

**Comment by aequitas** [comment]
@ericgj thanks for your response, I had a quick look at your code but didn't have enough time this week to really look into it.

@Bogdanp no problem, there is no hurry for me, at the moment, I have suspended the research for my active project to replace Celery with Dramatiq because I want to focus on some core functionality first. So I decided to work around the problems Celery is giving me now and try to apply Dramatiq at a later point for either fun or profit.

Looking at the implementation, which I think is remarkable clean and elegant, it seems to tick off everything.

- workflow creation and instantiation is isolated
- actors can receive results of other actors
- actors arguments can be set during workflow creation 
- a group of actors can be treated as a single entity

Also the DSL as shown in the testcase is clear and readable. I must confess I got used to Celery's abuse of the `|` operator (although deep in my heart I despise using the 'or' operator for that), but their alternative is to add the next task as a kwarg in the previous one. I think your solution here is much nicer. (after seeing your PR while writing this, I will confess know I'm glad you added the operator and I will use it as well ;)).

I slightly disagree with how the result of the previous task is passed as argument to the next. But this is due to Celery bias so you may completely ignore my opinion on this. Since Celery passes it as the first argument. Maybe a compromise would be to make reversing this a global configurable behaviour?

As for failures. I have adopted a mindset that tasks in chains never break the chain when failing. As in my opinion it is an easier model with one path with branching to much. For this I rely heavy on the retry concept.
A task now has 2 'failure' modes: it can 'fail' or produce an 'error' result. 

- A task 'fails' because of external conditions to the task context (temporary network issue, full disk, unexpected error in general) and these kind of failures always cause a task to be retried (on a different worker or time if needed).
- A task 'errors' because of a failure within the task context (expected errors). This error is then passed to the next task as the 'result'. Since most of the time the next task needs to do something with the result of the previous one (aggregating results or storing the result in the database) and a failure is also something that needs to be aggregated or stored.

This requires writing tasks in a different way and think more about things like idempotency, but I think that is worth the effort. 
Others might have their use cases for error handling of course.

I agree with the `group` and `pipeline` names, they are intuitive. `fanout` also gave me a short moment of confusion for the same reason @ericgj mentioned.

Many thanks and gratitude for your work in this short amount of time and completing it feature complete with tests and docs already. I created this Issue to hear your opinion about what direction Dramatiq would take on this matter but never expected to have a potential solution by the next weekend!

As I mentioned earlier I have put Dramatiq aside for now, but seeing the progress you made gives me confidence in this framework and the direction you have chosen for it and I will soon pick it up to consider as an alternative for Celery.
